src/coreclr/vm/threadsuspend.cpp

```diff
@@ -2099,104 +2091,146 @@ void Thread::RareDisablePreemptiveGC()
         goto Exit;
     }
 
+    _ASSERTE (m_fPreemptiveGCDisabled);
+
     // Holding a spin lock in preemp mode and switch to coop mode could cause other threads spinning
     // waiting for GC
     _ASSERTE ((m_StateNC & Thread::TSNC_OwnsSpinLock) == 0);
-
     _ASSERTE(!MethodDescBackpatchInfoTracker::IsLockOwnedByCurrentThread() || IsInForbidSuspendForDebuggerRegion());
 
     if (!GCHeapUtilities::IsGCHeapInitialized())
     {
         goto Exit;
     }
 
-    if (ThreadStore::HoldingThreadStore(this))
+    // A thread that performs GC may switch modes inside GC and come here.
+    // We will not try suspending a thread that is responsible for the suspension.
+    if (this == ThreadSuspend::GetSuspensionThread())
```

Comment:
> Is there some case where a thread is set as the suspension thread and it doesn't hold the TSL? I am wondering if we need this check when we've got the TSL check below.

The difference between these two checks is that this one is quite understandable - a thread that performs suspension certainly does not want itself being blocked. Places that perform mode switches (like inside GC) could specialcase the suspending thread, but it is easier to just handle the scenario here. 

The one below seems a bit more dangerous. That is a random thread holding TSL and trying to get into coop mode. If such scenario happens by accident and then the thread allocates, causes GC... etc, not sure what would happen.  
I initially had an assert instead of a check, but the assert was hit in tests, so I changed it to a condition. I am not happy about that. Maybe there is no way around this and some scenarios must do it (very carefully), but it seems fragile. 

You are right though - the next check that tests for TSL ownership subsumes the check for thread driving the suspension. I guess, I can remove the `this == ThreadSuspend::GetSuspensionThread()` check for now. 

I will try to follow up separately from this PR and see if we can avoid holding TSL into coop mode.

Reply:
I'm fine if you want to keep the check or convert it to an assert inside the TSL check. Either way comments might be nice to preserve the context you just described here in the code. Thanks!